### PR TITLE
Discourage multi-argument infix syntax: lint applications (`x op (a, b)`), also lint operator-name definitions

### PR DESCRIPTION
**Pull request by som-snytt**
Lint multiarg infix applications, for example `x op (a, b)`

Also lint method definitions with multiple parameters where the method name begins with an operator char, for example `def +(a: Int, b: Int)`, on the assumption that operator-style naming encourages infix at call site.

Message in both cases is "multiarg infix syntax looks like a tuple and will be deprecated"

This aligns Scala 2 with Dotty (and likely Scala 3) changes, as per https://github.com/lampepfl/dotty/pull/4311 and http://dotty.epfl.ch/docs/reference/changed-features/operators.html

### CODE REVIEW

**Comment by som-snytt** [comment]
Drafting to handle `(i, j) == (42, 17)`.

**Comment by som-snytt** [comment]
Also don't warn on deprecated definitions.

Or context bounds: `def ?[T: ru.TypeTag : ClassTag] = InternalInfo[T]`

File: src/compiler/scala/tools/nsc/ast/parser/Parsers.scala [hunk 1]
```diff
@@ -2796,6 +2799,12 @@ self =>
             }
             expr()
           }
+        if (settings.multiargInfix)
+          vparamss match {
+            case (_ :: _ :: _) :: Nil if settings.multiargInfix && Chars.isOperatorPart(name.decoded.last) =>
```
Comment (som-snytt) [inline src/compiler/scala/tools/nsc/ast/parser/Parsers.scala]: setting checked twice.

**Comment by som-snytt** [comment]
@lrytz it adds `-Wconf`!

**Review by lrytz — APPROVED** [review]
Nice work, thank you!

**Comment by som-snytt** [comment]
Thank-you, Lukas, now I can go to sleep.

File: src/compiler/scala/tools/nsc/ast/parser/Parsers.scala [hunk 2]
```diff
@@ -868,11 +867,18 @@ self =>
...
       }
-      def mkNamed(args: List[Tree]) = if (isExpr) args map treeInfo.assignmentToMaybeNamedArg else args
+      def mkNamed(args: List[Tree]) = if (isExpr) args.map(treeInfo.assignmentToMaybeNamedArg) else args
+      var isMultiarg = false
       val arguments = right match {
-        case Parens(Nil)  => literalUnit :: Nil
-        case Parens(args) => mkNamed(args)
-        case _            => right :: Nil
+        case Parens(Nil)               => literalUnit :: Nil
+        case Parens(args @ (_ :: Nil)) => mkNamed(args)
```
Comment (som-snytt) [inline src/compiler/scala/tools/nsc/ast/parser/Parsers.scala:874]: The inner parens are extraneous.

**Comment by SethTisue** [comment]
> where the method name is [...] a plain identifier ending in an operator character

@som-snytt I ran into this case today over in #8998 and it surprised me:

```
[error] /Users/tisue/scala.213/src/library/scala/sys/process/ProcessBuilderImpl.scala:139:9: multiarg infix syntax looks like a tuple and will be deprecated
[error]     def lineStream_!(log: ProcessLogger, capacity: Integer): Stream[String] = lineStream(withInput = false, nonZeroException = false, Some(log), capacity)
[error]         ^
```

at first I assumed this was just a bug, but now I see it was intentional. but I don't understand the design thinking here? `lineStream_!` doesn't seem like a good example of what you meant to discourage? what is a good example of what you mean to discourage?

**Comment by som-snytt** [comment]
@SethTisue I don't know, Seth, but I certainly meant to discourage `lineStream_!`.

It did occur to me to go looking in the ecosystem for usages of names with trailing ops. I was too lazy or something.

`sys.process` has a background in funky op names that history has not been kind to, but maybe there are other important use cases besides `setter_=`.

Edit: Seth is persuasive and convinced me that it shouldn't warn.

**Comment by jeffrey-aguilera** [comment]
This really does not like Slick. It appears that `scalac -X` does not have an entry for `-Xlint:multiarg-infix`

**Comment by som-snytt** [comment]
Multiarg infix applications are not behind the lint flag, only definitions of multiarg operators (which suggest they would be used infix).

It would be friendlier if applications were also behind the lint flag.

Edit: https://github.com/scala/bug/issues/12058

**Pull request merged**